docs/contributing/model/adding_a_new_model.md

```diff
@@ -0,0 +1,690 @@
+thinker_hidden_states = output.multimodal_output["24"]
+```
+
+### Key Points
+
+1. **Accessing Upstream Outputs**: Use `stage_list[source_stage_id].engine_outputs` to get outputs from the source stage
+2. **Extracting Data**: Access `output.multimodal_output[key]` to get specific hidden states or intermediate results
+   - Keys are defined by your model's `forward()` method when it creates `multimodal_outputs`
+3. **Device Management**: Move tensors to appropriate devices (CPU for serialization, GPU for processing)
+4. **Shape Transformations**: Reshape tensors as needed for the next stage (e.g., flattening codec codes)
+5. **Batch Handling**: Process each request in the batch separately and return a list
+
+### Complete Flow Diagram
+
+```
+┌─────────────────────────────────────────────────────────────┐
+│  Stage 0 (Thinker) Completes Processing                     │
+│  - Generates text tokens                                    │
+│  - Stores outputs via stage.set_engine_outputs()           │
+└──────────────────────┬──────────────────────────────────────┘
+                       │
+                       ▼
+┌─────────────────────────────────────────────────────────────┐
+│  Orchestrator (omni_llm.py)                                 │
+│  - Polls stage output queue                                 │
+│  - Detects stage completion                                 │
+│  - Calls next_stage.process_engine_inputs()                 │
+└──────────────────────┬──────────────────────────────────────┘
+                       │
+                       ▼
+┌─────────────────────────────────────────────────────────────┐
+│  OmniStage.process_engine_inputs()                          │
+│  - Checks if custom_process_input_func exists                │
+│  - Calls your transition function                           │
+└──────────────────────┬──────────────────────────────────────┘
+                       │
+                       ▼
+┌─────────────────────────────────────────────────────────────┐
+│  Your Transition Function                                   │
+│  (e.g., thinker2talker)                                     │
+│  - Extracts data from stage_list[0].engine_outputs         │
+│  - Transforms data format                                    │
+│  - Returns list[OmniTokensPrompt]                           │
+└──────────────────────┬──────────────────────────────────────┘
+                       │
+                       ▼
+┌─────────────────────────────────────────────────────────────┐
+│  Stage 1 (Talker) Receives Inputs                           │
+│  - Processes transformed inputs                              │
+│  - Generates codec codes                                    │
+└─────────────────────────────────────────────────────────────┘
+```
+
+### Implementation Example
+
+Create stage transition processors in `vllm_omni/model_executor/stage_input_processors/your_model_name.py`:
+
+```python
+# qwen3_omni.py
```

Comment:
your_model_name.py

btw, this example seems to too limited to qwen-omni architecture